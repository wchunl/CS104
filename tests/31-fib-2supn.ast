ROOT "" (0.0.0)
|   TOK_FUNCTION "" (6.22.1) function
|   |   TOK_TYPE_ID "" (6.22.1) typeid
|   |   |   TOK_VOID "void" (6.22.1) void
|   |   |   TOK_IDENT "fail" (6.22.6) variable
|   |   TOK_PARAM "(" (6.22.11) param
|   |   |   TOK_TYPE_ID "" (6.22.12) typeid
|   |   |   |   TOK_STRING "string" (6.22.12) string
|   |   |   |   TOK_IDENT "expr" (6.22.19) string variable (6.22.12)
|   |   |   TOK_TYPE_ID "" (6.22.25) typeid
|   |   |   |   TOK_STRING "string" (6.22.25) string
|   |   |   |   TOK_IDENT "file" (6.22.32) string variable (6.22.25)
|   |   |   TOK_TYPE_ID "" (6.22.38) typeid
|   |   |   |   TOK_INT "int" (6.22.38) int
|   |   |   |   TOK_IDENT "line" (6.22.42) int variable (6.22.38)
|   TOK_FUNCTION "" (6.24.1) function
|   |   TOK_TYPE_ID "" (6.24.1) typeid
|   |   |   TOK_VOID "void" (6.24.1) void
|   |   |   TOK_IDENT "putchr" (6.24.6) variable
|   |   TOK_PARAM "(" (6.24.13) param
|   |   |   TOK_TYPE_ID "" (6.24.14) typeid
|   |   |   |   TOK_INT "int" (6.24.14) int
|   |   |   |   TOK_IDENT "chr" (6.24.18) int variable (6.24.14)
|   TOK_FUNCTION "" (6.25.1) function
|   |   TOK_TYPE_ID "" (6.25.1) typeid
|   |   |   TOK_VOID "void" (6.25.1) void
|   |   |   TOK_IDENT "putint" (6.25.6) variable
|   |   TOK_PARAM "(" (6.25.13) param
|   |   |   TOK_TYPE_ID "" (6.25.14) typeid
|   |   |   |   TOK_INT "int" (6.25.14) int
|   |   |   |   TOK_IDENT "num" (6.25.18) int variable (6.25.14)
|   TOK_FUNCTION "" (6.26.1) function
|   |   TOK_TYPE_ID "" (6.26.1) typeid
|   |   |   TOK_VOID "void" (6.26.1) void
|   |   |   TOK_IDENT "putstr" (6.26.6) variable
|   |   TOK_PARAM "(" (6.26.13) param
|   |   |   TOK_TYPE_ID "" (6.26.14) typeid
|   |   |   |   TOK_STRING "string" (6.26.14) string
|   |   |   |   TOK_IDENT "str" (6.26.21) string variable (6.26.14)
|   TOK_FUNCTION "" (6.28.1) function
|   |   TOK_TYPE_ID "" (6.28.1) typeid
|   |   |   TOK_INT "int" (6.28.1) int
|   |   |   TOK_IDENT "getchr" (6.28.5) variable
|   |   TOK_PARAM "(" (6.28.11) param
|   TOK_FUNCTION "" (6.29.1) function
|   |   TOK_TYPE_ID "" (6.29.1) typeid
|   |   |   TOK_STRING "string" (6.29.1) string
|   |   |   TOK_IDENT "getstr" (6.29.8) variable
|   |   TOK_PARAM "(" (6.29.14) param
|   TOK_FUNCTION "" (6.30.1) function
|   |   TOK_TYPE_ID "" (6.30.1) typeid
|   |   |   TOK_STRING "string" (6.30.1) string
|   |   |   TOK_IDENT "getln" (6.30.8) variable
|   |   TOK_PARAM "(" (6.30.13) param
|   TOK_FUNCTION "" (7.8.1) function
|   |   TOK_TYPE_ID "" (7.8.1) typeid
|   |   |   TOK_INT "int" (7.8.1) int
|   |   |   TOK_IDENT "fibonacci" (7.8.5) variable
|   |   TOK_PARAM "(" (7.8.15) param
|   |   |   TOK_TYPE_ID "" (7.8.16) typeid
|   |   |   |   TOK_INT "int" (7.8.16) int
|   |   |   |   TOK_IDENT "n" (7.8.20) int variable (7.8.16)
|   |   TOK_BLOCK "{" (7.8.23)
|   |   |   TOK_IF "if" (7.9.4)
|   |   |   |   '<' "<" (7.9.10)
|   |   |   |   |   TOK_IDENT "n" (7.9.8) int variable (7.8.16)
|   |   |   |   |   TOK_INTCON "2" (7.9.12) int const
|   |   |   |   TOK_RETURN "return" (7.9.15)
|   |   |   |   |   TOK_IDENT "n" (7.9.22) int variable (7.8.16)
|   |   |   TOK_RETURN "return" (7.10.4)
|   |   |   |   '+' "+" (7.10.29) int vreg
|   |   |   |   |   TOK_CALL "(" (7.10.21)
|   |   |   |   |   |   TOK_IDENT "fibonacci" (7.10.11) variable
|   |   |   |   |   |   '-' "-" (7.10.24) int vreg
|   |   |   |   |   |   |   TOK_IDENT "n" (7.10.22) int variable (7.8.16)
|   |   |   |   |   |   |   TOK_INTCON "1" (7.10.26) int const
|   |   |   |   |   TOK_CALL "(" (7.10.41)
|   |   |   |   |   |   TOK_IDENT "fibonacci" (7.10.31) variable
|   |   |   |   |   |   '-' "-" (7.10.44) int vreg
|   |   |   |   |   |   |   TOK_IDENT "n" (7.10.42) int variable (7.8.16)
|   |   |   |   |   |   |   TOK_INTCON "2" (7.10.46) int const
|   TOK_FUNCTION "" (7.13.1) function
|   |   TOK_TYPE_ID "" (7.13.1) typeid
|   |   |   TOK_INT "int" (7.13.1) int
|   |   |   TOK_IDENT "main" (7.13.5) variable
|   |   TOK_PARAM "(" (7.13.9) param
|   |   TOK_BLOCK "{" (7.13.12)
|   |   |   TOK_VARDECL "=" (7.14.10) variable lval
|   |   |   |   TOK_INT "int" (7.14.4) int
|   |   |   |   TOK_IDENT "n" (7.14.8) int variable (7.8.16)
|   |   |   |   TOK_INTCON "0" (7.14.12) int const
|   |   |   TOK_WHILE "while" (7.15.4)
|   |   |   |   '<' "<" (7.15.13)
|   |   |   |   |   TOK_IDENT "n" (7.15.11) int variable (7.8.16)
|   |   |   |   |   TOK_INTCON "10" (7.15.15) int const
|   |   |   |   TOK_BLOCK "{" (7.15.19)
|   |   |   |   |   TOK_CALL "(" (7.16.14)
|   |   |   |   |   |   TOK_IDENT "putstr" (7.16.7) variable
|   |   |   |   |   |   TOK_STRINGCON ""fibonacci("" (7.16.15) string const
|   |   |   |   |   TOK_CALL "(" (7.17.14)
|   |   |   |   |   |   TOK_IDENT "putint" (7.17.7) variable
|   |   |   |   |   |   TOK_IDENT "n" (7.17.15) int variable (7.8.16)
|   |   |   |   |   TOK_CALL "(" (7.18.14)
|   |   |   |   |   |   TOK_IDENT "putstr" (7.18.7) variable
|   |   |   |   |   |   TOK_STRINGCON "") = "" (7.18.15) string const
|   |   |   |   |   TOK_CALL "(" (7.19.14)
|   |   |   |   |   |   TOK_IDENT "putint" (7.19.7) variable
|   |   |   |   |   |   TOK_CALL "(" (7.19.25)
|   |   |   |   |   |   |   TOK_IDENT "fibonacci" (7.19.15) variable
|   |   |   |   |   |   |   TOK_IDENT "n" (7.19.26) int variable (7.8.16)
|   |   |   |   |   TOK_CALL "(" (7.20.14)
|   |   |   |   |   |   TOK_IDENT "putchr" (7.20.7) variable
|   |   |   |   |   |   TOK_CHARCON "'\n'" (7.20.15) int const
|   |   |   |   |   '=' "=" (7.21.9) int vreg
|   |   |   |   |   |   TOK_IDENT "n" (7.21.7) int variable (7.8.16)
|   |   |   |   |   |   '+' "+" (7.21.13) int vreg
|   |   |   |   |   |   |   TOK_IDENT "n" (7.21.11) int variable (7.8.16)
|   |   |   |   |   |   |   TOK_INTCON "1" (7.21.15) int const
|   |   |   TOK_RETURN "return" (7.23.4)
|   |   |   |   TOK_INTCON "0" (7.23.11) int const
