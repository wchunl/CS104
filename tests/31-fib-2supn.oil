fail:     .function
          .param ptr expr
          .param ptr file
          .param int line
          .end
putchr:   .function
          .param int chr
          .end
putint:   .function
          .param int num
          .end
putstr:   .function
          .param ptr str
          .end
getchr:   .function int
          .end
getstr:   .function
          .end
getln:    .function
          .end
fibonacci:.function int
          .param int n
.if0:     $t0:i = n < 2
          goto .el0 if not $t0:i
.th0:     
.fi0:     return +
          return 
          .end
main:     .function int
          .local int n
          n = 0
.wh0:     $t1:i = n < 10
          goto .od0 if not $t1:i
.do0:     putstr("fibonacci(")
          putint(n)
          putstr(") = ")
          fibonacci(n)
          putint($t2:i)
          putchr('\n')
          $t3:i = n + 1
          n = $t3:i
.od0:     return 0
          return 
          .end
