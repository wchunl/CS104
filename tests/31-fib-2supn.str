string_set[   0]:    4706206694200543916 0x12e2dc8->","
                   16579916199511856337 0x12e4488->"getln"
string_set[   1]:    2297668033614959926 0x12e5b88->"0"
string_set[   3]:   11868120863286193964 0x12e5748->"main"
                    9611428142580219387 0x12de758->"fail"
string_set[   4]:      21606623943142546 0x12e3278->"{"
                    4017852018407011111 0x12e2ed8->"file"
string_set[   6]:   13686721134187162920 0x12e5138->"}"
string_set[   8]:    1370730858159036685 0x12e5398->"+"
                    9386170080689500517 0x12e2c78->"expr"
string_set[   9]:   10319632053171904911 0x12e5c38->"while"
string_set[  10]:    7419612306637001510 0x12e6158->"") = ""
string_set[  12]:   15816506932648033945 0x12e2ae8->"("
string_set[  14]:    9317282556992397945 0x12e4cb8->"if"
string_set[  17]:   18413295556186398326 0x12e32c8->"putint"
string_set[  18]:    7211130878751977441 0x12e3418->"putchr"
                   12679825168387489523 0x12e39a8->"num"
string_set[  23]:   11166701219533395393 0x12e41f8->"getstr"
                   12868445110721718657 0x12e2ef8->")"
string_set[  28]:   14635384904416252381 0x12e4f88->"return"
                   16378307159501682956 0x12e46f8->"fibonacci"
string_set[  29]:    2594088158685378223 0x12e5e58->"10"
                   12291392672316406109 0x12e30b8->"line"
string_set[  31]:    6192799629375248351 0x12e62b8->"'\n'"
                    6142509188972423790 0x12de738->""
                   13348923631108115385 0x12e3738->"putstr"
string_set[  32]:   11597697714117577063 0x12e5708->"="
string_set[  33]:    1583576014053361350 0x12e3b58->"str"
string_set[  34]:    7372855250431704371 0x12e5238->"-"
string_set[  38]:   11912410025211896923 0x12e5f08->""fibonacci(""
                   12597418120160561823 0x12e2cf8->"int"
string_set[  39]:    7296690267450279874 0x12e3588->"chr"
string_set[  40]:   11911796782241914524 0x12de318->"void"
string_set[  42]:    5344662657899890615 0x12e30d8->";"
string_set[  43]:    4551451650890805270 0x12e4ed8->"2"
string_set[  44]:   13497578857207247412 0x12e2b98->"string"
string_set[  45]:   10159970873491820195 0x12e52e8->"1"
                    1396094175042499165 0x12e4e28->"<"
                    9826907452181669771 0x12e4998->"n"
string_set[  46]:    9201800968376702500 0x12e3f88->"getchr"
load_factor = 0.851
bucket_count = 47
max_bucket_size = 3
