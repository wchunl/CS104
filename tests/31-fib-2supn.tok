#  1 "31-fib-2supn.oc"
#  1 "<built-in>"
#  1 "<command-line>"
#  1 "31-fib-2supn.oc"
#  1 "oclib.h"
# 22 "oclib.h"
   6 22.00001 261 TOK_VOID      void
   6 22.00006 282 TOK_IDENT     fail
   6 22.00011  40 '('           (
   6 22.00012 263 TOK_STRING    string
   6 22.00019 282 TOK_IDENT     expr
   6 22.00023  44 ','           ,
   6 22.00025 263 TOK_STRING    string
   6 22.00032 282 TOK_IDENT     file
   6 22.00036  44 ','           ,
   6 22.00038 262 TOK_INT       int
   6 22.00042 282 TOK_IDENT     line
   6 22.00046  41 ')'           )
   6 22.00047  59 ';'           ;
   6 24.00001 261 TOK_VOID      void
   6 24.00006 282 TOK_IDENT     putchr
   6 24.00013  40 '('           (
   6 24.00014 262 TOK_INT       int
   6 24.00018 282 TOK_IDENT     chr
   6 24.00021  41 ')'           )
   6 24.00022  59 ';'           ;
   6 25.00001 261 TOK_VOID      void
   6 25.00006 282 TOK_IDENT     putint
   6 25.00013  40 '('           (
   6 25.00014 262 TOK_INT       int
   6 25.00018 282 TOK_IDENT     num
   6 25.00021  41 ')'           )
   6 25.00022  59 ';'           ;
   6 26.00001 261 TOK_VOID      void
   6 26.00006 282 TOK_IDENT     putstr
   6 26.00013  40 '('           (
   6 26.00014 263 TOK_STRING    string
   6 26.00021 282 TOK_IDENT     str
   6 26.00024  41 ')'           )
   6 26.00025  59 ';'           ;
   6 28.00001 262 TOK_INT       int
   6 28.00005 282 TOK_IDENT     getchr
   6 28.00011  40 '('           (
   6 28.00012  41 ')'           )
   6 28.00013  59 ';'           ;
   6 29.00001 263 TOK_STRING    string
   6 29.00008 282 TOK_IDENT     getstr
   6 29.00014  40 '('           (
   6 29.00015  41 ')'           )
   6 29.00016  59 ';'           ;
   6 30.00001 263 TOK_STRING    string
   6 30.00008 282 TOK_IDENT     getln
   6 30.00013  40 '('           (
   6 30.00014  41 ')'           )
   6 30.00015  59 ';'           ;
#  7 "31-fib-2supn.oc"
   7 8.00001 262 TOK_INT       int
   7 8.00005 282 TOK_IDENT     fibonacci
   7 8.00015  40 '('           (
   7 8.00016 262 TOK_INT       int
   7 8.00020 282 TOK_IDENT     n
   7 8.00021  41 ')'           )
   7 8.00023 123 '{'           {
   7 9.00004 265 TOK_IF        if
   7 9.00007  40 '('           (
   7 9.00008 282 TOK_IDENT     n
   7 9.00010  60 '<'           <
   7 9.00012 283 TOK_INTCON    2
   7 9.00013  41 ')'           )
   7 9.00015 268 TOK_RETURN    return
   7 9.00022 282 TOK_IDENT     n
   7 9.00023  59 ';'           ;
   7 10.00004 268 TOK_RETURN    return
   7 10.00011 282 TOK_IDENT     fibonacci
   7 10.00021  40 '('           (
   7 10.00022 282 TOK_IDENT     n
   7 10.00024  45 '-'           -
   7 10.00026 283 TOK_INTCON    1
   7 10.00027  41 ')'           )
   7 10.00029  43 '+'           +
   7 10.00031 282 TOK_IDENT     fibonacci
   7 10.00041  40 '('           (
   7 10.00042 282 TOK_IDENT     n
   7 10.00044  45 '-'           -
   7 10.00046 283 TOK_INTCON    2
   7 10.00047  41 ')'           )
   7 10.00048  59 ';'           ;
   7 11.00001 125 '}'           }
   7 13.00001 262 TOK_INT       int
   7 13.00005 282 TOK_IDENT     main
   7 13.00009  40 '('           (
   7 13.00010  41 ')'           )
   7 13.00012 123 '{'           {
   7 14.00004 262 TOK_INT       int
   7 14.00008 282 TOK_IDENT     n
   7 14.00010  61 '='           =
   7 14.00012 283 TOK_INTCON    0
   7 14.00013  59 ';'           ;
   7 15.00004 267 TOK_WHILE     while
   7 15.00010  40 '('           (
   7 15.00011 282 TOK_IDENT     n
   7 15.00013  60 '<'           <
   7 15.00015 283 TOK_INTCON    10
   7 15.00017  41 ')'           )
   7 15.00019 123 '{'           {
   7 16.00007 282 TOK_IDENT     putstr
   7 16.00014  40 '('           (
   7 16.00015 285 TOK_STRINGCON "fibonacci("
   7 16.00027  41 ')'           )
   7 16.00028  59 ';'           ;
   7 17.00007 282 TOK_IDENT     putint
   7 17.00014  40 '('           (
   7 17.00015 282 TOK_IDENT     n
   7 17.00016  41 ')'           )
   7 17.00017  59 ';'           ;
   7 18.00007 282 TOK_IDENT     putstr
   7 18.00014  40 '('           (
   7 18.00015 285 TOK_STRINGCON ") = "
   7 18.00021  41 ')'           )
   7 18.00022  59 ';'           ;
   7 19.00007 282 TOK_IDENT     putint
   7 19.00014  40 '('           (
   7 19.00015 282 TOK_IDENT     fibonacci
   7 19.00025  40 '('           (
   7 19.00026 282 TOK_IDENT     n
   7 19.00027  41 ')'           )
   7 19.00028  41 ')'           )
   7 19.00029  59 ';'           ;
   7 20.00007 282 TOK_IDENT     putchr
   7 20.00014  40 '('           (
   7 20.00015 284 TOK_CHARCON   '\n'
   7 20.00019  41 ')'           )
   7 20.00020  59 ';'           ;
   7 21.00007 282 TOK_IDENT     n
   7 21.00009  61 '='           =
   7 21.00011 282 TOK_IDENT     n
   7 21.00013  43 '+'           +
   7 21.00015 283 TOK_INTCON    1
   7 21.00016  59 ';'           ;
   7 22.00004 125 '}'           }
   7 23.00004 268 TOK_RETURN    return
   7 23.00011 283 TOK_INTCON    0
   7 23.00012  59 ';'           ;
   7 24.00001 125 '}'           }
