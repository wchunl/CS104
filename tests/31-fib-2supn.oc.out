; 0.  1: # 1 "31-fib-2supn.oc"; 1.  0: 
; 1.  1: # 1 "<built-in>"
; 2.  1: # 1 "<command-line>"
; 3.  1: # 1 "31-fib-2supn.oc"
; 4.  1: 
; 4.  2: 
; 4.  3: 
; 4.  4: 
; 4.  5: 
; 4.  6: # 1 "oclib.h" 1
; 5.  1: # 22 "oclib.h"
; 6. 22: void fail (string expr, string file, int line);
; 6. 23: 
; 6. 24: void putchr (int chr);
; 6. 25: void putint (int num);
; 6. 26: void putstr (string str);
; 6. 27: 
; 6. 28: int getchr();
; 6. 29: string getstr();
; 6. 30: string getln();
; 6. 31: # 7 "31-fib-2supn.oc" 2
; 7.  7: 
; 7.  8: int fibonacci (int n) {
; 7.  9:    if (n < 2) return n;
; 7. 10:    return fibonacci (n - 1) + fibonacci (n - 2);
; 7. 11: }
; 7. 12: 
; 7. 13: int main() {
; 7. 14:    int n = 0;
; 7. 15:    while (n < 10) {
; 7. 16:       putstr ("fibonacci(");
; 7. 17:       putint (n);
; 7. 18:       putstr (") = ");
; 7. 19:       putint (fibonacci (n));
; 7. 20:       putchr ('\n');
; 7. 21:       n = n + 1;
; 7. 22:    }
; 7. 23:    return 0;
; 7. 24: }
